# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

In [2]:
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Видим, что в датафрейме всего 12 столбцов, в двух из которых есть пропущенные значения: в days_employed и total_income.
Датафрейм состоит из объектов,целых чисел и вещественных.

## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

*С помощью данных методов мы нашли где и сколько пропущенных значений: по 2174 в двух столбцах*

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

*С помощью метода abs() мы заменили отрицательные значения положительными в столбце days_employed*

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

*Вывели значения трудового стажа(в днях) для каждого типа занятости*

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

*Вывели уникальные значения столбца children и видим, что среди них есть некорректные*

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

*Удалим строки, в которых кол-во детей -1 или 20*

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

*Проверим данные в столбце после удаления: теперь все верно*

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

*Заполним пропуски с помощью цикла и логической индексации*

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

*Проверим успешность заполнения пропусков*

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

*Заменим значения типа float на int с помощью метода astype()*

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [15]:
data['education'] = data['education'].str.lower()

*Для исправления неявных дубликатов, приведем столбец education к нижнему регистру*

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

*Выведем кол-во дубликатов и удалим их*

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

*Создадим функцию для разделения доходов на категории*

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

*Применим ее к новому столбцу*

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

*Выведем уникальные значения столбца purpose методом unique(). \
Видим, что значений много, поскольку они повторяются. Их можно сократить*

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

*Создали функцию, применили ее к новому столбцу. Теперь в столбце purpose есть только 4 значения, так лучше с ними работать*

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Посмотрим на кол-во строк имевших задолженность по возврату по кол-ву детей. 

In [23]:
children_debt = data.groupby('children')['debt'].count()
children_debt# Ваш код будет здесь. Вы можете создавать новые ячейки.

children
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: debt, dtype: int64

Если я правильно поняла изначальные данные, то бездетные чаще берут кредиты, поскольку у них >14000 строк(среди которых люди, которые и имели задолженность и не имели)

In [24]:
children_have_debt = data.pivot_table(index = ['children'], \
                                columns = ['debt'], \
                                values = 'total_income', aggfunc='count')
children_have_debt['Общее кол-во заемщиков'] = data.groupby('children')['debt'].count()
children_have_debt['Средний % просрочки'] = children_have_debt[1]/ children_have_debt[0] *100
children_have_debt

debt            0       1  Общее кол-во заемщиков  Средний % просрочки
children                                                              
0         13028.0  1063.0                   14091             8.159349
1          4364.0   444.0                    4808            10.174152
2          1858.0   194.0                    2052            10.441335
3           303.0    27.0                     330             8.910891
4            37.0     4.0                      41            10.810811
5             9.0     NaN                       9                  NaN

**Вывод: Зависимость незначительна** \
Среди бездетных больше людей, имевших задолженность, но в соотношении у них процент должников ниже остальных. В целом, разница между количество детей и возвратом кредита в срок незначительна. \
В семьях с 5 детьми вообще нет должников. По ним мало данных для достоверной оценки и для корректности сравним три самые многочисленные выборки - семьи с количеством детей 0,1 или 2: \
Среди них кол-во должников растет с увеличением кол-ва детей. У бездетных процент должников ниже всех, между семьями с 1 и 2 ребенком разница небольшая, но она есть. Семьи с 1 ребенком чаще возвращают кредит в срок, чем семьи с 2 детьми.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [25]:
family_status_debt = data.groupby('family_status')['debt'].count()
family_status_debt# Ваш код будет здесь. Вы можете создавать новые ячейки.

family_status
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
гражданский брак          4134
женат / замужем          12261
Name: debt, dtype: int64

In [26]:
family_status_have_debt = data.pivot_table(index = ['family_status'], \
                                columns = ['debt'], \
                                values = 'total_income', aggfunc='count')
family_status_have_debt['Общее кол-во заемщиков'] = data.groupby('family_status')['debt'].count()
family_status_have_debt['Средний % просрочки'] = family_status_have_debt[1]/ family_status_have_debt[0] *100
family_status_have_debt

debt                       0    1  Общее кол-во заемщиков  Средний % просрочки
family_status                                                                 
Не женат / не замужем   2523  273                    2796            10.820452
в разводе               1105   84                    1189             7.601810
вдовец / вдова           888   63                     951             7.094595
гражданский брак        3749  385                    4134            10.269405
женат / замужем        11334  927                   12261             8.178931

**Вывод: Зависимость есть** \
В процентах самые низкие показатели у разведенных и вдов, т.е. вероятность возврата кредита у них выше. 
Незамужние и проживающие в гражданском браке хуже всех возвращают кредиты.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Посмотрим на минимальные,максимальные и медианные значения дохода

In [27]:
data['total_income'].min()# Ваш код будет здесь. Вы можете создавать новые ячейки.

20667

In [28]:
data['total_income'].max()

2265604

In [29]:
data['total_income'].median()

142594.0

Поделим их на 4 категории

In [30]:
def income(total_income):
    if total_income <= 90000:
        return 'Низкий уровень дохода'
    if total_income <= 150000:
        return 'Средний уровень дохода'
    if total_income <= 200000:
        return 'Высокий уровень дохода'
    return 'Очень высокий уровень дохода'

In [31]:
data['income_status'] = data['total_income'].apply(income)

In [32]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  \
0               покупка жилья                     B  операции с недвижимостью   
1     приобретение автомобиля                     C    операции с автомобилем   
2               покупка жилья                     C  операции с недвижимостью   
3  дополнительное образование                     B     получение образования   
4             сыграть свадьбу                     C        проведение свадьбы   

                  income_status  
0  Очень высокий уровень дохода  
1        Средний уровень дохода  
2        Средний уровень дохода  
3  Очень высокий уровень дохода  
4        Высокий уровень дохода

In [33]:
data['income_status'].value_counts()

Средний уровень дохода          8221
Очень высокий уровень дохода    5039
Высокий уровень дохода          4738
Низкий уровень дохода           3333
Name: income_status, dtype: int64

Большую часть составляют семьи со средним доходом, наименьшую - с низким.

In [34]:
income_have_debt = data.pivot_table(index = ['income_status'], \
                                columns = ['debt'], \
                                values = 'total_income', aggfunc='count')
income_have_debt['Общее кол-во заемщиков'] = data.groupby('income_status')['debt'].count()
income_have_debt['Средний % должников'] = income_have_debt[1]/ income_have_debt[0] *100
income_have_debt

debt                             0    1  Общее кол-во заемщиков  \
income_status                                                     
Высокий уровень дохода        4335  403                    4738   
Низкий уровень дохода         3071  262                    3333   
Очень высокий уровень дохода  4683  356                    5039   
Средний уровень дохода        7510  711                    8221   

debt                          Средний % должников  
income_status                                      
Высокий уровень дохода                   9.296424  
Низкий уровень дохода                    8.531423  
Очень высокий уровень дохода             7.601965  
Средний уровень дохода                   9.467377

**Вывод: Зависимость в возврате кредита в срок от дохода незначительна** \
Сравним из полученных данных две самые многочисленные выборки: семьи с средним уровнем дохода и очень высоким уровнем дохода. \
Среди семей с очень высоким уровнем дохода процент должников ниже и составляет 7,6 %, с средним уровнем дохода - он выше и составляет 9,5%. 

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [35]:
purpose_have_debt = data.pivot_table(index = ['purpose_category'], \
                                columns = ['debt'], \
                                values = 'total_income', aggfunc='count')
purpose_have_debt['Общее кол-во заемщиков'] = data.groupby('purpose_category')['debt'].count()
purpose_have_debt['Средний % должников'] = purpose_have_debt[1]/purpose_have_debt[0] * 100
purpose_have_debt# Ваш код будет здесь. Вы можете создавать новые ячейки.

debt                         0    1  Общее кол-во заемщиков  \
purpose_category                                              
операции с автомобилем    3879  400                    4279   
операции с недвижимостью  9971  780                   10751   
получение образования     3619  369                    3988   
проведение свадьбы        2130  183                    2313   

debt                      Средний % должников  
purpose_category                               
операции с автомобилем              10.311936  
операции с недвижимостью             7.822686  
получение образования               10.196187  
проведение свадьбы                   8.591549

**Вывод: Зависимость между целью кредита и возврата его в срок есть** \
Лучше возвращают кредит на операции с недвижимостью, думаю, это связано с тем, что к данной цели люди подходят более осознанно и если, например, это ипотека, то люди не могут не платить за нее, т.к. многие платят ипотеку и в этой квартире и живут, т.е. своевременная невыплата может грозить более серьезными последствиями и ежемесячный платеж по ипотеке - регулярное обязательство.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Это может быть сбой в системе (как при внесении, так и при передаче данных), человеческий фактор (например, вместо "15" человек может написать "150")

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* В случае если в данных есть некорректные или большие/маленькие значения, медиана сможет "не реагировать" на них и выдать реальное значение, в отличии от средней, на которую эти значения могут сильно повляить

### Шаг 4: общий вывод.

Работа была проведена с информацией о кредитозаемщиках, с помощью которой были выявлены зависимости о возврате кредита в срок между количеством детей, семейным положением, уровнем дохода и целями кредита. \
На этапе предобработки данных выявили пропущенные значения в столбцах и заменили их медианными, заменили/удалили аномальные значения, заменили в необходимых столбцах вещественные типы данных на целочисленные, выявили и обработали дубликаты, а также поделили некоторые столбцы на категории.  

В целом, зависимость наблюдается. \
Между кол-вом детей: надежнее всего бездетные семьи, среди них наименьший процент должников - 8,2 %.\
Между семейным положением: самый низкий процент должников среди разведенных(7,6%) и вдов(7,1 %).\
Между уровнем дохода: семьи с очень высоким уровнем дохода(200000 тыс. руб +) и средним уровнем дохода(150000 тыс. руб +) лучше возвращают кредиты в срок.\
Между целями кредита: самыми возвращаемыми в срок являются кредиты на недвижимость, хуже всего возвращают автокредиты и кредиты на образование.

Рекомендации для банка: 
- Для подсчета трудового стажа использовать годы, а не дни.
- Установить диапазон вводимых значений, например, на кол-во детей во избежание аномальных значений.
- При заполнении анкеты указать рекомендации для клиентов,например, использование целых значений в столбце "ежемесячный доход" и использование единого регистра.
- В столбце "цель кредита" сделать список с несколькими категориями, чтобы люди выбирали подходящий под них, а не вводили сами.